app/src/main/java/com/example/cinemaddict/ui/discover/DiscoverFragment.kt

```diff
@@ -1,8 +1,121 @@
 import dagger.hilt.android.AndroidEntryPoint
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.launch
 
 @AndroidEntryPoint
-class DiscoverFragment : BaseUiFragment<FragmentDiscoverBinding>(FragmentDiscoverBinding::inflate)
\ No newline at end of file
+class DiscoverFragment : BaseUiFragment<FragmentDiscoverBinding>(FragmentDiscoverBinding::inflate) {
+
+    private lateinit var adapter: GenrePagerAdapter
+    override val viewModel: DiscoverViewModel by viewModels()
+
+    private val movieAdapter: MovPagingAdapter<FilmDiscoverData, ItemDiscoverScreenBinding> by lazy {
+        MovPagingAdapter(
+            bindingInflater = ItemDiscoverScreenBinding::inflate,
+            onClickListener = ::navigateToDetailScreen
+        )
+    }
+    private val stateListener: LoadStateListener<FilmDiscoverData, ItemDiscoverScreenBinding> by lazy {
+        LoadStateListener(
+            adapter = movieAdapter,
+            toastCallback = ::toast
+        )
+    }
+
+    override fun initViews() {
+        super.initViews()
+        binding.viewModel = viewModel
+        binding.rvList.apply {
+            layoutManager = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
+            adapter = movieAdapter
+        }
+    }
+
+    override fun initListeners() {
+        super.initListeners()
+        onRefresh(Dispatchers.Main) {
+            viewModel.refresh()
+        }
+
+        stateListener.isLoad.observe(viewLifecycleOwner){
+            if (it) showLoader() else hideLoader()
+        }
+    }
+
+    @OptIn(FlowPreview::class)
+    override fun initObservers() {
+        super.initObservers()
+        viewModel.genre.observe(viewLifecycleOwner) {
+            adapter = GenrePagerAdapter(childFragmentManager, lifecycle)
+            adapter.setGenres(it)
+            binding.vpPager.adapter = adapter
+
+            binding.vpPager.offscreenPageLimit = 1
+            TabLayoutMediator(
+                binding.tlGenre,
+                binding.vpPager,
+                adapter
+            ).attach()
+        }
+
+        viewModel.error.observe(viewLifecycleOwner) {
+            toast(it)
+        }
+
+        viewModel.searchText.observe(viewLifecycleOwner) {
+            if (!it.isNullOrEmpty()) {
+                isGetMovie(true)
+                viewModel.getMovie(it)
+            } else {
+                isGetMovie(false)
+                viewModel.getMovie("")
+            }
+        }
+
+        viewLifecycleOwner.lifecycleScope.launch {
+            viewModel.dataFilm.debounce(300)
+                .distinctUntilChanged()
+                .collectLatest { list ->
+                    movieAdapter.submitData(list)
+                }
+        }
+    }
+
+    private fun isGetMovie(isGet: Boolean) {
```

Comment:
it's not a correct way to name functions like this (it's more common for variables)
functions must describe action (for example, set***Visibility(isContainQuery)